xdpderp/xdpderp.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package xdpderp
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"net"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -type config -type stats_key -type stats_per_af bpf xdp.c -- -I headers
+
+// STUNServer manages loading and unloading of an eBPF XDP program that serves
+// the STUN protocol. It exports statistics for the XDP program via its
+// implementation of the prometheus.Collector interface.
+type STUNServer struct {
+	mu    sync.Mutex
+	objs  bpfObjects
+	stats *stunServerStats
+}
+
+// STUNServerConfig represents the configuration of a STUNServer.
+type STUNServerConfig struct {
+	DeviceName  string
+	DstPort     int
+	AttachFlags link.XDPAttachFlags
+}
+
+func (s *STUNServerConfig) validate() error {
+	if len(s.DeviceName) < 1 {
+		return errors.New("DeviceName is unspecified")
+	}
+	if s.DstPort < 0 || s.DstPort > math.MaxUint16 {
+		return errors.New("DstPort is outside of uint16 bounds")
+	}
+	return nil
+}
+
+// NewSTUNServer returns an instance of a STUNServer that has attached the STUN
+// XDP program to the netdev and destination port specified by config.
+func NewSTUNServer(config *STUNServerConfig) (*STUNServer, error) {
+	err := config.validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid config: %v", err)
+	}
+	iface, err := net.InterfaceByName(config.DeviceName)
+	if err != nil {
+		return nil, fmt.Errorf("error finding device: %w", err)
+	}
+	objs := bpfObjects{}
+	err = loadBpfObjects(&objs, &ebpf.CollectionOptions{
+		Programs: ebpf.ProgramOptions{
+			LogSize: ebpf.DefaultVerifierLogSize * 10,
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error loading XDP program: %w", err)
+	}
+	server := &STUNServer{
+		objs: objs,
+	}
+	var key uint32
+	xdpConfig := bpfConfig{
+		DstPort: uint16(config.DstPort),
+	}
+	err = objs.ConfigMap.Put(key, &xdpConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error loading config in eBPF map: %w", err)
+	}
+	_, err = link.AttachXDP(link.XDPOptions{
+		Program:   objs.XdpProgFunc,
+		Interface: iface.Index,
+		Flags:     config.AttachFlags,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error attaching XDP program to dev: %w", err)
+	}
+	server.stats = newSTUNServerStats()
+	return server, nil
+}
+
+// Close unloads the XDP program and associated maps.
+func (s *STUNServer) Close() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.objs.Close()
+}
+
+type stunServerStats struct {
+	registry *prometheus.Registry
+	// These are effectively counters, but we declare them as gauges so that we
+	// can set absolute values instead of being limited to increment ops. The
+	// values being set are coming out of an eBPF map that only sees increment
+	// ops, but we don't track changes.
```

Comment:
I couldn't find concrete documentation on how this would affect Prometheus server downstream in the ways you're describing. I'm not convinced the promql rate functions that handle wraps care about the underlying metric type, or if it's even propagated via the HELP text when scraped, but I haven't tested. Generally speaking I want them to "fit" into the counter type. I don't believe we can read and zero the map values atomically, so storing the last value is probably a better approach.

Reply:
Does Prometheus just store the value then? It probably also stores the data type, and I guess it's up to something like Grafana to do the interpretation?

Admittedly I was extrapolating my understanding based on how things like rrdtool handle gauge vs counter.

Reply:
Not sure. Rather than dig into those implementation details I just made this use native prometheus counters in https://github.com/tailscale/tailscale/pull/12390/commits/5ec490556ae55d012c8276de824aafd06b8083b5.